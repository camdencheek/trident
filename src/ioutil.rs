use anyhow::Result;
use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
use std::fs::File;
use std::io::{Read, Seek, SeekFrom, Write};
use std::marker::PhantomData;
use std::os::unix::fs::FileExt;
use std::sync::Arc;

use crate::build::serialize::StreamWriter;

#[derive(Clone)]
pub struct FileCursor {
    f: Arc<File>,
    offset: u64,
}

impl FileCursor {
    pub fn new(f: File) -> Self {
        Self {
            f: Arc::new(f),
            offset: 0,
        }
    }
}

impl Read for FileCursor {
    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
        let n = self.f.read_at(buf, self.offset)?;
        self.offset += n as u64;
        Ok(n)
    }

    fn read_exact(&mut self, buf: &mut [u8]) -> std::io::Result<()> {
        self.f.read_exact_at(buf, self.offset)?;
        self.offset += buf.len() as u64;
        Ok(())
    }
}

// impl Write for FileCursor {
//     fn write(&mut self, buf: &[u8]) -> Result<usize> {
//         let n = self.f.write_at(buf, self.offset)?;
//         self.offset += n as u64;
//         Ok(n)
//     }

//     fn write_all(&mut self, buf: &[u8]) -> Result<()> {
//         self.f.write_all_at(buf, self.offset)?;
//         self.offset += buf.len() as u64;
//         Ok(())
//     }

//     fn flush(&mut self) -> Result<()> {
//         self.f.flush()
//     }
// }

impl Seek for FileCursor {
    fn seek(&mut self, pos: SeekFrom) -> std::io::Result<u64> {
        match pos {
            SeekFrom::Current(i) => self.offset = (self.offset as i64 + i) as u64,
            SeekFrom::Start(i) => self.offset = i,
            SeekFrom::End(i) => self.offset = (self.f.metadata()?.len() as i64 + i) as u64,
        };
        Ok(self.offset)
    }
}

pub trait SectionType {}
impl<T: SectionType> SectionType for Section<T> {}
impl SectionType for () {}

#[derive(Debug, Copy, Clone)]
pub struct Section<P: SectionType = ()> {
    pub offset: u64,
    pub len: u64,
    _parent_type: PhantomData<P>,
}

impl<T: SectionType> StreamWriter for Section<T> {
    fn write_to<W: Write>(&self, w: &mut W) -> Result<usize> {
        w.write_u64::<LittleEndian>(self.offset)?;
        w.write_u64::<LittleEndian>(self.len)?;
        Ok(std::mem::size_of::<u64> as usize * 2)
    }
}

impl<P: SectionType> Section<P> {
    pub fn new(offset: u64, len: u64) -> Self {
        Self {
            offset,
            len,
            _parent_type: PhantomData::<P>,
        }
    }

    pub fn narrow(&self, child: Section<Self>) -> Self {
        assert!(child.offset + child.len <= self.len);
        Self::new(self.offset + child.offset, child.len)
    }
}
