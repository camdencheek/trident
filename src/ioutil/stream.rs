use anyhow::Result;
use std::io::{self, Read, Write};

use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};

pub trait StreamWrite {
    fn write_to<W: Write>(&self, w: &mut W) -> io::Result<usize>;

    fn to_vec(&self) -> Vec<u8> {
        let mut v = Vec::new();
        self.write_to(&mut v)
            .expect("write_to should not fail if the underlying writer does not return an error");
        v
    }
}

impl StreamWrite for u8 {
    fn write_to<W: Write>(&self, w: &mut W) -> io::Result<usize> {
        w.write_u8(*self)?;
        Ok(1)
    }
}

impl StreamWrite for u16 {
    fn write_to<W: Write>(&self, w: &mut W) -> io::Result<usize> {
        w.write_u16::<BigEndian>(*self)?;
        Ok(2)
    }
}

impl StreamWrite for u32 {
    fn write_to<W: Write>(&self, w: &mut W) -> io::Result<usize> {
        w.write_u32::<BigEndian>(*self)?;
        Ok(4)
    }
}

impl StreamWrite for u64 {
    fn write_to<W: Write>(&self, w: &mut W) -> io::Result<usize> {
        w.write_u64::<BigEndian>(*self)?;
        Ok(8)
    }
}

impl<const N: usize> StreamWrite for [u8; N] {
    fn write_to<W: Write>(&self, w: &mut W) -> io::Result<usize> {
        w.write(self)?;
        Ok(N)
    }
}

pub trait StreamRead
where
    Self: Sized,
{
    fn read_from<R: Read>(r: &mut R) -> Result<Self>;
}

impl StreamRead for u8 {
    fn read_from<R: Read>(r: &mut R) -> Result<Self> {
        Ok(r.read_u8()?)
    }
}

impl StreamRead for u16 {
    fn read_from<R: Read>(r: &mut R) -> Result<Self> {
        Ok(r.read_u16::<BigEndian>()?)
    }
}

impl StreamRead for u32 {
    fn read_from<R: Read>(r: &mut R) -> Result<Self> {
        Ok(r.read_u32::<BigEndian>()?)
    }
}

impl StreamRead for u64 {
    fn read_from<R: Read>(r: &mut R) -> Result<Self> {
        Ok(r.read_u64::<BigEndian>()?)
    }
}

impl<const N: usize> StreamRead for [u8; N] {
    fn read_from<R: Read>(r: &mut R) -> Result<Self> {
        let mut buf = [0u8; N];
        r.read(&mut buf)?;
        Ok(buf)
    }
}
