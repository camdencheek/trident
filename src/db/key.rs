use anyhow::anyhow;
use std::io::{self, Read, Write};

use crate::ioutil::stream::{StreamRead, StreamWrite};

// TODO should these be defined in a higher-level module?
type Trigram = [u8; 3];
type PartitionID = u16;
type OID = [u8; 20];
type BlockID = u32;
type DocID = u32;

#[derive(PartialEq, Eq, Clone, Debug)]
pub enum DBKey {
    Partition(PartitionID, PartitionKey),
}

impl DBKey {
    fn discriminant(&self) -> u8 {
        match self {
            Self::Partition(_, _) => 0,
        }
    }
}

impl StreamWrite for DBKey {
    fn write_to<W: Write>(&self, w: &mut W) -> io::Result<usize> {
        let mut n = self.discriminant().write_to(w)?;
        match self {
            Self::Partition(id, key) => {
                n += id.write_to(w)?;
                n += key.write_to(w)?;
            }
        }
        Ok(n)
    }
}

impl StreamRead for DBKey {
    fn read_from<R: Read>(r: &mut R) -> anyhow::Result<Self> {
        match u8::read_from(r)? {
            0 => Ok(Self::Partition(
                PartitionID::read_from(r)?,
                PartitionKey::read_from(r)?,
            )),
            _ => Err(anyhow!("bad discriminant")),
        }
    }
}

#[derive(PartialEq, Eq, Clone, Debug)]
pub enum PartitionKey {
    Index(IndexKey),
    Contents(DocID),
}

impl PartitionKey {
    fn discriminant(&self) -> u8 {
        match self {
            Self::Index(_) => 0,
            Self::Contents(_) => 1,
        }
    }
}

impl StreamWrite for PartitionKey {
    fn write_to<W: Write>(&self, w: &mut W) -> io::Result<usize> {
        let mut n = self.discriminant().write_to(w)?;
        match self {
            Self::Index(key) => n += key.write_to(w)?,
            Self::Contents(oid) => n += oid.write_to(w)?,
        };
        Ok(n)
    }
}

impl StreamRead for PartitionKey {
    fn read_from<R: Read>(r: &mut R) -> anyhow::Result<Self> {
        match u8::read_from(r)? {
            0 => Ok(Self::Index(IndexKey::read_from(r)?)),
            1 => Ok(Self::Contents(DocID::read_from(r)?)),
            _ => Err(anyhow!("bad discriminant")),
        }
    }
}

#[derive(PartialEq, Eq, Clone, Debug)]
pub enum IndexKey {
    TrigramPosting(Trigram, TrigramPostingKey),
}

impl IndexKey {
    fn discriminant(&self) -> u8 {
        match self {
            Self::TrigramPosting(_, _) => 0,
        }
    }
}

impl StreamWrite for IndexKey {
    fn write_to<W: Write>(&self, w: &mut W) -> io::Result<usize> {
        let mut n = self.discriminant().write_to(w)?;
        match self {
            Self::TrigramPosting(trigram, key) => {
                n += trigram.write_to(w)?;
                n += key.write_to(w)?;
            }
        };
        Ok(n)
    }
}

impl StreamRead for IndexKey {
    fn read_from<R: Read>(r: &mut R) -> anyhow::Result<Self> {
        match u8::read_from(r)? {
            0 => Ok(Self::TrigramPosting(
                Trigram::read_from(r)?,
                TrigramPostingKey::read_from(r)?,
            )),
            _ => Err(anyhow!("bad discriminant")),
        }
    }
}

#[derive(PartialEq, Eq, Clone, Debug)]
pub enum TrigramPostingKey {
    SuccessorCount,
    MatrixCount,
    DocCount,
    SuccessorsBlock(BlockID),
    MatrixBlock(BlockID),
    DocsBlock(BlockID),
}

impl TrigramPostingKey {
    fn discriminant(&self) -> u8 {
        match self {
            Self::SuccessorCount => 0,
            Self::MatrixCount => 1,
            Self::DocCount => 2,
            Self::SuccessorsBlock(_) => 3,
            Self::MatrixBlock(_) => 4,
            Self::DocsBlock(_) => 5,
        }
    }
}

impl StreamWrite for TrigramPostingKey {
    fn write_to<W: Write>(&self, w: &mut W) -> io::Result<usize> {
        let mut n = self.discriminant().write_to(w)?;
        match self {
            Self::SuccessorCount | Self::MatrixCount | Self::DocCount => {}
            Self::SuccessorsBlock(b) | Self::MatrixBlock(b) | Self::DocsBlock(b) => {
                n += b.write_to(w)?
            }
        };
        Ok(n)
    }
}

impl StreamRead for TrigramPostingKey {
    fn read_from<R: Read>(r: &mut R) -> anyhow::Result<Self> {
        match u8::read_from(r)? {
            0 => Ok(Self::SuccessorCount),
            1 => Ok(Self::MatrixCount),
            2 => Ok(Self::DocCount),
            3 => Ok(Self::SuccessorsBlock(BlockID::read_from(r)?)),
            4 => Ok(Self::MatrixBlock(BlockID::read_from(r)?)),
            5 => Ok(Self::DocsBlock(BlockID::read_from(r)?)),
            _ => Err(anyhow!("bad discriminant")),
        }
    }
}

#[cfg(test)]
mod test {
    use std::io::Cursor;

    use super::*;
    use insta::assert_debug_snapshot;
    use quickcheck::{quickcheck, Arbitrary};

    impl Arbitrary for DBKey {
        fn arbitrary(g: &mut quickcheck::Gen) -> Self {
            Self::Partition(PartitionID::arbitrary(g), PartitionKey::arbitrary(g))
        }
    }

    impl Arbitrary for PartitionKey {
        fn arbitrary(g: &mut quickcheck::Gen) -> Self {
            match u8::arbitrary(g) % 2 {
                0 => Self::Index(IndexKey::arbitrary(g)),
                1 => Self::Contents(DocID::arbitrary(g)),
                _ => unreachable!(),
            }
        }
    }

    impl Arbitrary for IndexKey {
        fn arbitrary(g: &mut quickcheck::Gen) -> Self {
            Self::TrigramPosting(Trigram::arbitrary(g), TrigramPostingKey::arbitrary(g))
        }
    }

    impl Arbitrary for TrigramPostingKey {
        fn arbitrary(g: &mut quickcheck::Gen) -> Self {
            match u8::arbitrary(g) % 6 {
                0 => Self::SuccessorCount,
                1 => Self::MatrixCount,
                2 => Self::DocCount,
                3 => Self::SuccessorsBlock(BlockID::arbitrary(g)),
                4 => Self::MatrixBlock(BlockID::arbitrary(g)),
                5 => Self::DocsBlock(BlockID::arbitrary(g)),
                _ => unreachable!(),
            }
        }
    }

    quickcheck! {
        // Test that any DBKey can be roundtripped
        fn db_key_roundtrip(key: DBKey) -> bool {
            let v = key.to_vec();
            let mut r = Cursor::new(v);
            key == DBKey::read_from(&mut r).unwrap()
        }
    }

    // This test checks that the sort order of the set of keys does not change.
    // This is important for iterating over entries in order.
    #[test]
    fn stable_sort_order() {
        let keys = [
            DBKey::Partition(
                42,
                PartitionKey::Index(IndexKey::TrigramPosting(
                    *b"abc",
                    TrigramPostingKey::DocCount,
                )),
            ),
            DBKey::Partition(
                42,
                PartitionKey::Index(IndexKey::TrigramPosting(
                    *b"abc",
                    TrigramPostingKey::MatrixBlock(24),
                )),
            ),
            DBKey::Partition(
                42,
                PartitionKey::Index(IndexKey::TrigramPosting(
                    *b"abc",
                    TrigramPostingKey::SuccessorCount,
                )),
            ),
            DBKey::Partition(
                42,
                PartitionKey::Index(IndexKey::TrigramPosting(
                    *b"abc",
                    TrigramPostingKey::MatrixBlock(42),
                )),
            ),
            DBKey::Partition(42, PartitionKey::Contents(0)),
            DBKey::Partition(42, PartitionKey::Contents(2)),
            DBKey::Partition(
                35,
                PartitionKey::Index(IndexKey::TrigramPosting(
                    *b"abc",
                    TrigramPostingKey::DocCount,
                )),
            ),
        ];

        let mut serialized = keys.iter().map(|key| key.to_vec()).collect::<Vec<_>>();
        serialized.sort();

        assert_debug_snapshot!(serialized)
    }
}
