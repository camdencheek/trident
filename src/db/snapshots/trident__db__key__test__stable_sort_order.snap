---
source: src/db/key.rs
expression: serialized
---
[
    [
        0,
        0,
        35,
        0,
        0,
        97,
        98,
        99,
        2,
    ],
    [
        0,
        0,
        42,
        0,
        0,
        97,
        98,
        99,
        0,
    ],
    [
        0,
        0,
        42,
        0,
        0,
        97,
        98,
        99,
        2,
    ],
    [
        0,
        0,
        42,
        0,
        0,
        97,
        98,
        99,
        4,
        0,
        0,
        0,
        24,
    ],
    [
        0,
        0,
        42,
        0,
        0,
        97,
        98,
        99,
        4,
        0,
        0,
        0,
        42,
    ],
    [
        0,
        0,
        42,
        1,
        0,
        0,
        0,
        0,
    ],
    [
        0,
        0,
        42,
        1,
        0,
        0,
        0,
        2,
    ],
]
